tla-pp/src/test/scala/at/forsyte/apalache/tla/pp/TestQuantificationOptimizer.scala

```diff
@@ -0,0 +1,104 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir.{Typed, _}
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.transformations.impl.IdleTracker
+import org.junit.runner.RunWith
+import at.forsyte.apalache.tla.lir.TypedPredefs._
+import at.forsyte.apalache.tla.lir.oper.{TlaArithOper, TlaBoolOper}
+import at.forsyte.apalache.tla.lir.values.{TlaIntSet, TlaNatSet}
+import org.scalacheck.Gen
+import org.scalacheck.Prop.{forAll, AnyOperators}
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatestplus.junit.JUnitRunner
+import org.scalatestplus.scalacheck.Checkers
+
+@RunWith(classOf[JUnitRunner])
+class TestQuantificationOptimizer extends AnyFunSuite with Checkers {
+
+  val optimizer = new IntegerQuantificationOptimizer(new IdleTracker)
+
+  val gens: IrGenerators = new IrGenerators {
+    override val maxArgs: Int = 3
+  }
+
+  val ops =
+    gens.simpleOperators ++ gens.logicOperators ++ gens.arithOperators ++ gens.setOperators ++ gens.functionOperators
+
+  val rangeGen: Gen[TlaEx] = for {
+    a <- gens.genTlaEx(ops)(gens.emptyContext)
+    b <- gens.genTlaEx(ops)(gens.emptyContext)
+  } yield tla.dotdot(a.as(IntT1()), b.as(IntT1())).as(SetT1(IntT1()))
+
+  val boolTag = Typed(BoolT1())
+  val varname = tla.name("x").as(IntT1())
+
+  def forSetWithPredicates(
+      setGen: Gen[TlaEx],
+      postPredE: (TlaEx, TlaEx) => TlaEx,
+      postPredA: (TlaEx, TlaEx) => TlaEx): Unit = {
+    val pairGen = for {
+      setEx <- setGen
+      exRaw <- gens.genTlaEx(ops)(gens.emptyContext)
+    } yield (setEx, exRaw)
+    val prop = forAll(pairGen) { case (setEx, exRaw) =>
+      // gens produces garbage w.r.t. type tags, we just force the toplevel ex to be boolean.
+      val ex = exRaw.withTag(boolTag)
+      val existsEx = tla.exists(varname, setEx, ex).as(BoolT1())
+      val forallEx = tla.forall(varname, setEx, ex).as(BoolT1())
+
+      val existsCond = postPredE(setEx, ex)
+      val forallCond = postPredA(setEx, ex)
```

Comment:
Neither `{exists,forall}Cond` nor `postPred{A,E}` tell me that these are the _expected_ conditions. I think they should be renamed to something more descriptive.